lib/kernel/test/os_SUITE.erl

```diff
@@ -469,6 +470,19 @@ error_info(Config) ->
         ],
     error_info_lib:test_error_info(os, L).
 
+os_cmd_shell(Config) ->
+    DataDir = proplists:get_value(data_dir, Config),
+    SysShell = filename:join(DataDir, "sys_shell"),
+
+    {ok, OldShell} = application:get_env(kernel, os_cmd_shell),
+    application:set_env(kernel, os_cmd_shell, SysShell),
+
+    %% os:cmd should not try to detect the shell location rather than use
+    %% the value from kernel:os_cmd_shell parameter
+    Ls = os:cmd("ls"),
+    application:set_env(kernel, os_cmd_shell, OldShell),
+    comp("sys_shell", Ls).
```

Comment:
```suggestion
    try
        application:set_env(kernel, os_cmd_shell, SysShell),

        %% os:cmd should not try to detect the shell location rather than use
        %% the value from kernel:os_cmd_shell parameter
        comp("sys_shell", os:cmd("ls"))
    after
        application:set_env(kernel, os_cmd_shell, OldShell),
    end.
```

I've found that this is generally a better way to make sure that the old state is restored.